Data-cube ready to use

floglass opened the pull request:
Working fine if aperture size < 1000, otherwise _very_ slow.
I personally use apertures of 101 pixels.

The output data-cube contains 11 layers, the first one is the polychromatic psf, and then each of the wavelengths used to create the polychrome.
If no aperture.fits file in the directory, the code will generate a 101x101 pixels HST-like aperture.

5 verbose levels added. Accessible with the option -v when calling the code. The verbose levels are: debug, info, warning, error, critical.
The default level is info. If no information is wanted, I suggest using -v critical.

floglass committed:
Polychromatic PSF saved as data-cube

But the creation is very slow (>10 minutes with aperture of 1666x1666).

floglass committed:
Size of array properly passed throughout the code.

For both monochrome or polychrome

floglass committed:
verbose level added to reduce console output

using logging module.

floglass committed:
negative values in output PSF fixed

no need to prefilter the data with spline_filter. Option in scipy..geometric_transform([...],  prefilter=False)
Output looks correct now.

floglass committed:
minor bugfix

Was calling poly.dist when instancing the object, but it is actually computed later in create_polychrome().

floglass committed:
debug output added and consistency checked.

.fits output are consistent with release 2.0
output is now a data cube containing the polychrome, and the 10 wavelengths.

floglass committed:
Polychromatic PSF in data-cube ready

Still uses distortion.par file for distortions, but not for long!

ijiraq commented on modules.py:
```diff
@@ -52,10 +60,27 @@ def dist(self, file_path='distortions.par'):
+        else:
+            hdu = pyfits.PrimaryHDU()
+            list_arrays = list()
+            list_arrays.append(self.a)
```
I had expected that the data cube would be in one extension and the image of the polychromatic PSF would be in another?  Then the user can choose to either just use the 'POLYPSF' extension or the 'MONOPSF' one?

ijiraq commented on modules.py:
```diff
@@ -83,10 +108,10 @@ def save(self, name=None):
 
         if name is None:
```
This code structure could be simplified by a line like:

name = "{}.fits".format(name is not None and name or self.name)
hdu.writeto('name, clobber=True)
print >> sys.stdout, 'Saving {}'.format(name)

and then just one 'save' chunk is needed.  The above line sets the value of the variable 'name' to the value of the variable 'name' if the current value of 'name' is not None other-wise name=self.name and then

ijiraq commented on modules.py:
```diff
@@ -203,6 +230,7 @@ def fits2aperture(self):
 
         self.array_size = np.shape(self.a)[0]
         self.name = 'aperture'
+        logging.info("'aperture.fits' loaded, image size=%s" % self.array_size)
```
this logging message should be updated given the 'input_file' variable value instead of the fixed value 'aperture.fits'

ijiraq commented on modules.py:
```diff
@@ -210,8 +238,10 @@ class Pupil(OpticalArray):
+    init takes: wavelength, array_size=1666
     """
-    def __init__(self, wavelength, array_size=101):
+    def __init__(self, wavelength, array_size=1666):
```
since 1666 is not reasonable perhaps this should default to a smaller value?  Is the default value here consistent with the default aperture.fits file?

ijiraq commented on modules.py:
```diff
@@ -219,18 +249,17 @@ def __init__(self, wavelength, array_size=101):
         aperture = Aperture(self.array_size)
         if glob.glob('aperture.fits'):
-            print "... using 'aperture.fits'..."
+            logging.info("... using 'aperture.fits'...")
```
is this the name of the fits file always? I thought you were switching to a variable name.

ijiraq commented on modules.py:
```diff
@@ -357,13 +400,22 @@ def wavelength_contributions(self):
+            logging.debug("pupil array_size=%s" % pupil.array_size)
+            psf = PSF(pupil, self.scale, self.array_size)
+            psf.resize_psf()  # scale is supposed to be 0.01
+            if logging.debug:
```
Interesting, I didn't know logging.debug evaluated to a boolean in these cases.

floglass committed:
saving code simplified

to a single line.

floglass committed:
logging.info() uses input_file

instead of hard coded 'aperture.fits'
Ijiraq comment # 3 addressed

floglass committed:
default size removed.

The size of the PSF is determined by the size of the aperture. If no aperture.fits is present in the directory, a new 101x101 pixel HST-like aperture will be created.

floglass committed:
fixed logging.debug

logging.debug does not return a boolean.
logging.getLogger().getEffectiveLevel() == 'debug' check if root logger is at level 'debug'

floglass committed:
new option -a for aperture

Can load any .fits aperture file as aperture

floglass committed:
polychrome stored outside the individuals PSFs extension

The data-cube now only contains the 10 individual PSFs used to make the polychrome. The polychrome is stored in a second extension in the .fits file.
A switch (-i) has been added to save all the individual PSFs in their own .fits file (for debugging purpose).

floglass committed:
comment removed

floglass commented:
The code creates a two extensions file: the first one contains all the monochromatic PSFs in a data-cube, and the second extension contains the polychromatic PSF.
The option -a allows for changing the name of the aperture file to use (if no file "aperture.fits" is present in the directory, and if no option is provided, an HST-like aperture will be created).
The size of the output PSF is determined by the size of the aperture mask file. The PSF will be five time bigger than the aperture.
